mountpoint-s3/src/checksums.rs

```diff
@@ -63,47 +67,77 @@ impl ChecksummedBytes {
         }
     }
 
-    /// Append the given checksummed bytes to current `ChecksummedBytes`, ensure that data integrity will
-    /// be validated.
+    /// Returns a slice of self for the provided range.
+    ///
+    /// This operation just increases the reference count and sets a few indices,
+    /// so there will be no validation and the checksum will not be recomputed.
+    pub fn slice(&self, range: impl RangeBounds<usize>) -> Self {
+        Self {
+            orig_bytes: self.orig_bytes.clone(),
+            curr_slice: self.curr_slice.slice(range),
+            checksum: self.checksum,
+        }
+    }
+
+    /// Returns a copy of this slice, with the guarantee that the checksum is computed exactly
+    /// on the slice, rather than on a larger containing buffer.
+    ///
+    /// Return [IntegrityError] if data corruption is detected.
+    pub fn shrink_to_fit(&self) -> Result<Self, IntegrityError> {
+        if self.curr_slice.len() == self.orig_bytes.len() {
+            return Ok(self.clone());
+        }
+
+        let result = Self::from_bytes(self.curr_slice.clone());
+        self.validate()?;
```

Comment:
This is actually a little subtle — it's only correct because `Self::from_bytes(self.curr_slice.clone())` doesn't copy the underlying bytes, and so `self.validate()` is checking the same physical bytes that the new checksum was computed over in `from_bytes`. If there was a copy involved, we'd have to do some extra work to make sure that copy had its checksum computed at the right time.

I think I'd rather not call `from_bytes` here and instead just do the construction manually, so we can be explicit that this transformation is safe because it's not copying. (It's very unlikely but not impossible that some future developer comes along and changes `from_bytes` to copy, without realizing some callers are making a bracketing assumption). We should probably also add a comment on the `curr_bytes` field to capture an invariant that it's always a subslice of `orig_bytes`.

Reply:
Thanks for highlighting this, missed this as a rather big assumption.

Agree, I would prefer to do the construction manually.

Reply:
>We should probably also add a comment on the curr_bytes field to capture an invariant that it's always a subslice of orig_bytes.

Is there some way we can codify that invariant? For example, not by using `split_off` but instead taking a (real) slice of `orig_bytes` and tying the lifetimes together?

Reply:
Self-referential lifetimes are pretty annoying (usually require `unsafe`), and `Bytes` is pretty heavily designed around reference counting, so probably not something we can codify that way.

If we wanted an executable check, we could do something janky like:
```rust
let curr_slice = &self.curr_slice[..];
// will panic if curr_slice isn't a subslice of orig_slice
let _ = self.orig_slice.slice_ref(curr_slice);
```
maybe as a debug assert? But possibly not worth the trouble.

Reply:
Got it. Plus even if the lifetimes worked, I don't think it would even be `Send` anymore and unusable in async code.

Random idea: maybe we consider storing the bounds/range in place of `curr_slice`, and create the smaller `Bytes` at the last minute when performing `into_bytes`. Looks like an involved refactor though at first glance.

No opinion here, I'm OK keeping it as a documented invariant.

Reply:
I've added the comment to `curr_slice` for now